cli/ssh.go

```diff
@@ -795,3 +805,104 @@ func remoteGPGAgentSocket(sshClient *gossh.Client) (string, error) {
 
 	return string(bytes.TrimSpace(remoteSocket)), nil
 }
+
+type closerWithName struct {
+	name   string
+	closer io.Closer
+}
+
+type closerStack struct {
+	sync.Mutex
+	closers []closerWithName
+	closed  bool
+	logger  slog.Logger
+	err     error
+}
+
+func newCloserStack(ctx context.Context, logger slog.Logger) *closerStack {
+	cs := &closerStack{logger: logger}
+	go cs.closeAfterContext(ctx)
+	return cs
+}
+
+func (c *closerStack) closeAfterContext(ctx context.Context) {
+	<-ctx.Done()
+	c.close(ctx.Err())
+}
+
+func (c *closerStack) close(err error) {
+	c.Lock()
+	if c.closed {
+		c.Unlock()
+		return
+	}
+	c.closed = true
+	c.err = err
+	c.Unlock()
+
+	for i := len(c.closers) - 1; i >= 0; i-- {
+		cwn := c.closers[i]
+		cErr := cwn.closer.Close()
+		c.logger.Debug(context.Background(),
+			"closed item from stack", slog.F("name", cwn.name), slog.Error(cErr))
+	}
```

Comment:
You don't know a priori whether a failure high in the stack will affect things lower in the stack, so you run all the closers.  We just log any failures.  The idea here is that it's a bit like a `defer` stack, but with some extra features.

`ssh` kicks off a `waitAndClose()` goroutine that monitors the context and whether the workspace gets stopped, and then closes stuff.  Also, this goroutine gets kicked off _before_ we know everything we want to close (in particular, before we process remote forwarding).  So, I built this so that:

a) you can add stuff to be closed later
b) you can close it from a goroutine OR a `defer` and it will only close stuff once.